### PR TITLE
add query ability to both list function and model

### LINKED ISSUE #37: Add query string support to shopcart List function
**As a** User  
 **I need** to search the shopcart with shopcart_id or customer_id
 **So that** I can know the details of the target shopcart
   
 ### Details and Assumptions
 * User can search the shopcart with shopcart_id, curstomer_id, or both.
   
 ### Acceptance Criteria  
   
 ```gherkin
 Given the shopcart(id=1, customer_id=15) exists
 When I call 'GET /shopcarts?id=1&customer_id=15'
 Then I will get the information of the shopcart 
 ```

### PR DESCRIPTION
**Pull request by huwwww**

### CODE REVIEW

File: tests/test_models.py [hunk 1]
```diff
@@ -244,4 +244,29 @@ def test_delete_shopcart_item(self):
...
+
+    def test_find_by_shopcart_id_and_customer_id(self):
+        """It should Find Shopcarts by shopcart_id and customer_id"""
+        shopcart = ShopcartFactory()
+        shopcart.create()
+
+        same_shopcart = Shopcart.find_by_shopcart_id_and_customer_id(shopcart.id, shopcart.customer_id)[0]
+        self.assertEqual(shopcart.id, same_shopcart.id)
+        self.assertEqual(shopcart.customer_id, same_shopcart.customer_id)
+
```
Comment (allenpthuang) [inline tests/test_models.py:258]: Do we need a sad path here? Just my 2 ct :)

File: tests/test_models.py [hunk 2]
```diff
@@ -244,4 +244,29 @@ def test_delete_shopcart_item(self):
...
+    def test_find_by_shopcart_id(self):
+        """It should Find Shopcarts by shopcart_id"""
+        shopcart = ShopcartFactory()
+        shopcart.create()
+
+        same_shopcart = Shopcart.find_by_shopcart_id(shopcart.id)[0]
+        self.assertEqual(shopcart.id, same_shopcart.id)
+
```
Comment (allenpthuang) [inline tests/test_models.py:266]: ditto

File: tests/test_models.py [hunk 3]
```diff
@@ -244,4 +244,29 @@ def test_delete_shopcart_item(self):
...
+    def test_find_by_customer_id(self):
+        """It should Find Shopcarts by customer_id"""
+        shopcart = ShopcartFactory()
+        shopcart.create()
+
+        same_shopcart = Shopcart.find_by_customer_id(shopcart.customer_id)[0]
+        self.assertEqual(shopcart.customer_id, same_shopcart.customer_id)
```
Comment (allenpthuang) [inline tests/test_models.py]: ditto

File: service/routes.py [hunk 4]
```diff
@@ -134,9 +134,26 @@ def list_all_shopcarts():
...
+    elif arg_customer_id:
+        for sc in shopcart.find_by_customer_id(arg_customer_id):
+            shopcarts["shopcarts"].append(sc.serialize())
+            app.logger.info(f"Retrieved shopcarts with specific customer_id: {shopcarts}")
+    else:
+        for sc in shopcart.all():
+            shopcarts["shopcarts"].append(sc.serialize())
+            app.logger.info(f"Retrieved shopcarts: {shopcarts}")
 
     return shopcarts, status.HTTP_200_OK
```
Comment (allenpthuang) [inline service/routes.py:175]: Do we need to return HTTP 404 if no such Shopcart exists? Currently, it returns an empty list of Shopcarts with HTTP 200. Just my 2ct, thanks!

File: tests/test_routes.py [hunk 5]
```diff
@@ -321,3 +322,57 @@ def test_read_item(self):
             f"{BASE_URL}/{shopcart_id}/items/{non_existing_item_id}"
         )
         self.assertEqual(resp.status_code, status.HTTP_404_NOT_FOUND)
+
+    def test_query_shopcarts_by_shopcart_id_and_customer_id(self):
```
Comment (allenpthuang) [inline tests/test_routes.py:327]: Ditto on sad paths :)

**Review by allenpthuang — COMMENTED** [review]
Thanks, Weiqi! I left some comments, but they are just my opinions. Feel free to let me know if I missed something! (e.g. returning HTTP 404 isn't required in #37 but I just think it should have been?) Thanks again!

File: tests/test_models.py (hunk 1 above)
Comment (huwwww) [reply tests/test_models.py:258]: Sure I will add it, thank you for your suggestion!

Comment (allenpthuang) [reply tests/test_models.py:258]: Thanks!

File: tests/test_models.py (hunk 2 above)
Comment (allenpthuang) [reply tests/test_models.py:266]: Thanks!

File: tests/test_models.py (hunk 3 above)
Comment (allenpthuang) [reply tests/test_models.py]: Thanks!

**Review by allenpthuang — APPROVED** [review]
Thanks, Weiqi!

**Review by yuheng-lu — APPROVED** [review]

**Review by sweta1947 — APPROVED** [review]

**Pull request merged**